File: Week 02/id_118/Python/LeetCode_77_118.py
from typing import List


class Solution:
    """
    res is C(n,k)
    Follow strictly the generic recursion template
    """

    def __init__(self):
        self.res = []

    def _gen_comb(self, n, k, start, curr_c):
        # terminator
        if len(curr_c) == k:
            self.res.append(curr_c.copy())
            return
        # process the current level logic
        for i in range(start, n + 1):
            curr_c.append(i)
            # drill down
            self._gen_comb(n, k, i + 1, curr_c)

            # reverse to reset for the next recursion
            curr_c.pop()
        return

    def combine(self, n: int, k: int) -> List[List[int]]:
        # edge cases
        if n <= 0 or k <= 0 or k > n:
            return self.res
        curr_c = []
        self._gen_comb(n, k, 1, curr_c)
        return self.res


class Solution2:
    """
    A more optimized solution by removing unnecessary branches
    """

    def __init__(self):
        self.res = []

    def _gen_comb(self, n, k, start, curr_c):
        # terminator
        if len(curr_c) == k:
            self.res.append(curr_c.copy())
            return
        # process the current level logic
        # here no need to go to all the way to n (n+1 is not included in py)
        # for any given state, the available slots are k - len(curr_c)
        # so the stop point is n + 1 - (k - len(curr_c)) + 1
        for i in range(start, n - (k - len(curr_c)) + 2):
            curr_c.append(i)
            # drill down
            self._gen_comb(n, k, i + 1, curr_c)

            # reverse to reset for the next recursion
            curr_c.pop()
        return

    def combine(self, n: int, k: int) -> List[List[int]]:
        # edge cases
        if n <= 0 or k <= 0 or k > n:
            return self.res
        curr_c = []
        self._gen_comb(n, k, 1, curr_c)
        return self.res


if __name__ == '__main__':
    sol = Solution()
    res = sol.combine(7, 4)
    print(res)
    sol2 = Solution2()
    res2 = sol2.combine(7, 4)
    print(res2)

File: Week 02/id_118/C++/LeetCode_77_118.cpp
//
// Created by Isaac Zhou on 2019-10-26.
//

#include <iostream>
#include <vector>

using namespace std;


class Solution {

private:
    vector <vector<int>> res;

    // To solve C(n,k)
    // c stores the current solution
    // start to begin searching new element
    void genComb(int n, int k, int start, vector<int> &c) {
        // terminator
        if (c.size() == k) {
            res.push_back(c);
            return;
        }
        // process the current level
        for (int i = start; i <= n; ++i) {
            c.push_back(i);
            // drill down
            genComb(n, k, i + 1, c);
            // reverse and reset for the next recursion
            c.pop_back();
        }
        return;
    }

public:
    vector <vector<int>> combine(int n, int k) {
        res.clear();
        //edge cases: meaningless cases
        if (n <= 0 || k <= 0 || k > n) {
            return res;
        }
        vector<int> c;
        genComb(n, k, 1, c);
        return res;
    }
};


class Solution2 {
/*
 * This solution is a more optimized one, it removed the unnecessary branches
 * */
private:
    vector <vector<int>> res;

    // To solve C(n,k)
    // c stores the current solution
    // start to begin searching new element
    void genComb(int n, int k, int start, vector<int> &c) {
        // terminator
        if (c.size() == k) {
            res.push_back(c);
            return;
        }
        // process the current level
        // here's an optimized tweak, no need to go to all the way to n
        // for the current level, there are only k - c.size() positions to be filled
        // hence the stop point is n - (k-c.size()) + 1
        for (int i = start; i <= n - (k-c.size()) + 1; ++i) {
            c.push_back(i);
            // drill down
            genComb(n, k, i + 1, c);
            // reverse and reset for the next recursion
            c.pop_back();
        }
        return;
    }

public:
    vector <vector<int>> combine(int n, int k) {
        res.clear();
        //edge cases: meaningless cases
        if (n <= 0 || k <= 0 || k > n) {
            return res;
        }
        vector<int> c;
        genComb(n, k, 1, c);
        return res;
    }
};

int main() {
    vector <vector<int>> res = Solution().combine(7, 4);
    for (int i = 0; i < res.size(); ++i) {
        cout << "Solution 1" <<endl;
        for (int j = 0; j < res[i].size(); ++j) {
            cout << res[i][j] << " ";
        }
        cout << endl;
    }

    vector <vector<int>> res = Solution2().combine(7, 4);
    for (int i = 0; i < res.size(); ++i) {
        cout << "Solution 2" <<endl;
        for (int j = 0; j < res[i].size(); ++j) {
            cout << res[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}